File: aab-web-application/_js/networking/eventHandling/jsonRpcCommunication.js
import {ClientConnectionType} from "../webSocket/clientConnectionType.js";

/**
 * Defines the direction in which a message is send.
 * @type {{from: string, to: string}}
 * @enum
 */
const JsonRpcMessageDirection = {
    "to": "to",
    "from": "from"
}

/**
 * Represnts the json rpcs communication information.
 */
class JsonRpcCommunication {
    //#region Properties
    /**
     * @type string
     * @private
     */
    _direction = undefined;
    /**
     * @return {string}
     */
    get Direction() { return this._direction; }
    /**
     * @type string
     * @private
     */
    _name = undefined;
    /**
     * @return {string}
     */
    get Name() { return this._name; }
    /**
     * @type string
     * @private
     */
    _clientConnectionType = ClientConnectionType.NotDefined;
    /**
     * @return {string}
     */
    get ClientConnectionType() { return this._clientConnectionType; }
    //#endregion

    /**
     * Creates a json rpc communication object based on a given jObject.
     * @param {{direction:string, name:string, clientConnectionType:string}} jsonDict - The json object representation of the data needed for such an object.
     * @return {null|JsonRpcCommunication} - The deserialized object, or null.
     * @static
     */
    static FromDict(jsonDict) {
        if (!(jsonDict.direction in JsonRpcMessageDirection))
            return null;

        let communication = new JsonRpcCommunication();

        communication._direction = jsonDict.direction;
        communication._name = jsonDict.name;
        communication._clientConnectionType = jsonDict.clientConnectionType;

        return communication;
    }

    /**
     * Creates a communication object that has as the direction "to".
     * @param {string} name - The name of the connection to communicate to.
     * @param {string} ccType - The type of the connection to communicate to.
     * @return {JsonRpcCommunication} - A new communication.
     * @static
     */
    static CommunicationTo({name, ccType}) {
        let communication = new JsonRpcCommunication();

        communication._direction = JsonRpcMessageDirection.to;
        communication._name = name;
        communication._clientConnectionType = ccType;

        return communication;
    }

    /**
     * Creates a communication object that has as the direction "from".
     * @param {string} name - The name of the connection this communication comes from..
     * @param {string} ccType - The type of the connection this communication comes from.
     * @return {JsonRpcCommunication} - A new communication.
     * @static
     */
    static CommunicationFrom({name, ccType}) {
        let communication = new JsonRpcCommunication();

        communication._direction = JsonRpcMessageDirection.from;
        communication._name = name;
        communication._clientConnectionType = ccType;

        return communication;
    }

    /**
     * Transforms this communication object into an json dict.
     * @return {{clientConnectionType: string, name: string, direction: string}} - The json dict representation.
     */
    ToJsonDict() {
        return {
            "direction": this._direction,
            "name": this._name,
            "clientConnectionType": this._clientConnectionType
        }
    }
}

export {
    JsonRpcCommunication,
    JsonRpcMessageDirection
}

File: aab-hl-application/Assets/00 BasicCode/1 Script/Networking/EventHandling/JsonRpcCommunication.cs
using MasterNetworking.WebSocket;
using Newtonsoft.Json.Linq;
using System;

namespace MasterNetworking.EventHandling
{
    /// <summary>
    /// Represnts the json rpcs communication information.
    /// </summary>
    public class JsonRpcCommunication
    {
        /// <summary>
        /// The direction this message is meant for.
        /// Either "to" or "from".
        /// </summary>
        public string Direction { get; private set; }
        /// <summary>
        /// The name the message should be send "to" is comming "from".
        /// </summary>
        public string Name { get; private set; } = null;
        /// <summary>
        /// The device type this message should be send "to" is comming "from".
        /// </summary>
        public ClientConnectionType? ClientConnectionType { get; private set; } = null;

        /// <summary>
        /// Creates a json rpc communication object based on a given jObject.
        /// </summary>
        /// <param name="jObject">The jObject representation of the data needed for such an object.</param>
        /// <returns>The deserialized object, or null.</returns>
        public static JsonRpcCommunication FromJObject(JObject jObject)
        {
            try
            {
                JsonRpcCommunication communication = new JsonRpcCommunication();
                communication.Direction = jObject["direction"].ToString();

                // Checks if the direction is correctly specefied.
                if (communication.Direction != "from" && communication.Direction != "to")
                    return null;

                if (jObject["name"] != null)
                    communication.Name = jObject["name"].ToString();
                if (jObject["clientConnectionType"] != null)
                    communication.ClientConnectionType = ClientConnectionTypeUtils.FromString(jObject["clientConnectionType"].ToString()).Value;

                return communication;
            }
            catch (Exception ex)
            {
                return null;
            }

        }

        /// <summary>
        /// Creates a communication object that has as the direction "to".
        /// </summary>
        /// <param name="name">The name of the connection to communicate to.</param>
        /// <param name="ccType">The type of the connection to communicate to.</param>
        /// <returns>A new communication.</returns>
        public static JsonRpcCommunication CommunicationTo(string name, ClientConnectionType? ccType)
        {
            return new JsonRpcCommunication()
            {
                Direction = "to",
                Name = name,
                ClientConnectionType = ccType
            };
        }

        /// <summary>
        /// Creates a communication object that has as the direction "from".
        /// </summary>
        /// <param name="name">The name of the connection this communication comes from.</param>
        /// <param name="ccType">The type of the connection this communication comes from.</param>
        /// <returns>A new communication.</returns>
        public static JsonRpcCommunication CommunicationFrom(string name, ClientConnectionType? ccType)
        {
            return new JsonRpcCommunication()
            {
                Direction = "from",
                Name = name,
                ClientConnectionType = ccType
            };
        }

        /// <summary>
        /// Transforms this commmunication object into an JObjects.
        /// </summary>
        /// <returns>The JObject representation.</returns>
        public JObject ToJObject()
        {
            string ccType = null;
            if (ClientConnectionType.HasValue)
                ccType = ClientConnectionTypeUtils.ToString(ClientConnectionType.Value);

            return new JObject(
                new JProperty("direction", Direction),
                new JProperty("name", Name),
                new JProperty("clientConnectionType", ccType)
            );
        }
    }
}
